### DISCUSSION TITLE
[Forking Renovate Bot] Dependency Lookup Warning with `ghcr`

### DISCUSSION THREAD

**Discussion by mpeddada1** [renovatebot/renovate#20534] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Thanks for an amazing project, team!

We're experiencing an issue where Forking Renovate is unable to look up dependencies hosted on `gchr.io`. On the other hand, the same dependency is successfully found and updated by Mend Renovate. 

Result of **Forking Renovate:**
<img width="969" alt="Screenshot 2023-02-17 at 4 39 47 PM" src="https://user-images.githubusercontent.com/66699525/219798952-02ab87f0-5f68-4189-8489-38d515478d9f.png">

Result of **Mend Renovate**
<img width="929" alt="Screenshot 2023-02-17 at 4 33 45 PM" src="https://user-images.githubusercontent.com/66699525/219798201-f7bb5ea9-5989-4d94-ba41-47687624ea2a.png">

However, when provided with a Maven dependency and an `ubuntu` image hosted on Docker Hub, both Forking Renovate and Mend Renovate are able to look them up successfully.  Any insights that you could provide into why there is a discrepancy in behavior for the two variations of renovate bot would be really helpful, thank you!

Here is the reproducer : https://github.com/mpeddada1/forking-renovate-repro

### Relevant debug logs

https://gist.github.com/mpeddada1/263ed189762bdc83b1086ea739a0bc40

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
Quick questions:
- Is it a private package on GHCR?
- Why does your log seem to show it _working_? https://gist.github.com/mpeddada1/263ed189762bdc83b1086ea739a0bc40#file-logs-for-forking-renovate-repro-L261-L277

**Comment by viceice** [comment]
@rarkins i recently saw some autoclosed PR for public ghcr images too. maybe some temporary network issues? i use the normal hosted app

**Comment by rarkins** [comment]
I just checked and neither app requests this permission currently:
![image](https://user-images.githubusercontent.com/6311784/220308872-8c2a434f-9de8-4402-b1a5-e8b0dc700739.png)

However it shouldn't be necessary in this case.

When I do a dry run on the reproduction repo it succeeds and I get these logs:

```
         "https://ghcr.io/token (GET,200)": 1,
         "https://ghcr.io/v2/ (GET,401)": 1,
         "https://ghcr.io/v2/graalvm/jdk/blobs/sha256:0be8f5a95274128d05fdbe113f950844596232ace4214820257d7ed681f1a1c9 (GET,200)": 1,
         "https://ghcr.io/v2/graalvm/jdk/manifests/latest (GET,200)": 1,
         "https://ghcr.io/v2/graalvm/jdk/manifests/sha256:0880a79cfd39495186e303e4a1dd928e5996909dcfae461c50146fd05097cd6f (GET,200)": 1,
         "https://ghcr.io/v2/graalvm/jdk/tags/list (GET,200)": 1,
         "https://ghcr.io/v2/graalvm/jdk/tags/list (GET,401)": 1,
```

If I duplicate the repo and install into Forking Renovate, I see:

```
    "https://ghcr.io/token (GET,403)": 1,
    "https://ghcr.io/v2/graalvm/jdk/tags/list (GET,401)": 1,
```

```
DEBUG: GET https://ghcr.io/token?service=ghcr.io&scope=repository:graalvm/jdk:pull = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=107)
DEBUG: Not allowed to access docker registry
{
  "registryHost": "https://ghcr.io"
  "dockerRepository": "graalvm/jdk"
}

DEBUG: Response code 403 (Forbidden)
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1676973295352,
      "socket": 1676973295352,
      "lookup": 1676973295394,
      "connect": 1676973295397,
      "secureConnect": 1676973295401,
      "upload": 1676973295401,
      "response": 1676973295459,
      "end": 1676973295459,
      "phases": {
        "wait": 0,
        "dns": 42,
        "tcp": 3,
        "tls": 4,
        "request": 0,
        "firstByte": 58,
        "download": 0,
        "total": 107
      }
    },
    "message": "Response code 403 (Forbidden)",
    "stack": "HTTPError: Response code 403 (Forbidden)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/34.125.1/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/34.125.1 (https://github.com/renovatebot/renovate)",
        "accept": "application/json",
        "authorization": "***********",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://ghcr.io/token?service=ghcr.io&scope=repository:graalvm/jdk:pull",
      "hostType": "docker",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 403,
      "statusMessage": "Forbidden",
      "body": {
        "errors": [
          {
            "code": "DENIED",
            "message": "denied"
          }
        ]
      },
      "headers": {
        "content-type": "application/json",
        "docker-distribution-api-version": "registry/2.0",
        "date": "Tue, 21 Feb 2023 09:54:55 GMT",
        "content-length": "50",
        "x-github-request-id": "D995:1DBD:359F42:50CCAF:63F494EF",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
}

DEBUG: Failed to get authHeaders for getTags lookup
DEBUG: Failed to look up docker dependency ghcr.io/graalvm/jdk
{
  "dependency": "ghcr.io/graalvm/jdk"
  "packageFile": "sample_docker.cfg"
}

```

**Comment by viceice** [comment]
do you add a hostrule? I saw we had that but reverted it later

- #19490

**Comment by rarkins** [comment]
Forking Renovate now runs the "full" Renovate slim image as if you were using the CLI. The main Renovate app still has the old backend with lots of integration points. Neither have any special rules for `ghcr.io`.

My suspicion was that Forking Renovate is sending the installation token with its requests to https://ghcr.io but considering we reverted that above feature, I'm not sure why that would be.

I have now added Packages permissions to Forking Renovate and accepted them. I saw no immediate fix, although perhaps there's caching in both Renovate and GitHub and either could be causing old tokens or permissions to be applied.

**Comment by rarkins** [comment, marked as answer]
I found the problem. The Forking Renovate app applies a host rule for hostType=docker to avoid `docker.io` rate limiting, but it's missing a `matchHost` setting so gets mistakenly sent to `ghcr.io` too. We will fix this server side
